packages/data-service/src/graphqlObservable.ts

```diff
@@ -0,0 +1,309 @@
+import { Observable } from "rxjs/Observable";
+import "rxjs/add/observable/throw";
+import "rxjs/add/observable/of";
+import "rxjs/add/observable/fromPromise";
+import "rxjs/add/operator/combineLatest";
+import "rxjs/add/operator/do";
+import "rxjs/add/operator/map";
+import "rxjs/add/operator/concatMap";
+
+import {
+  DefinitionNode,
+  DocumentNode,
+  getNamedType,
+  GraphQLInterfaceType,
+  GraphQLObjectType,
+  GraphQLSchema,
+  GraphQLType,
+  SelectionNode,
+  FieldNode,
+  GraphQLField,
+  GraphQLFieldResolver,
+  isTypeSystemDefinitionNode,
+  isTypeSystemExtensionNode,
+  Kind
+} from "graphql";
+
+// WARNING: This is NOT a spec complete graphql implementation
+// https://facebook.github.io/graphql/October2016/
+
+interface TypeMap {
+  [key: string]: GraphQLType;
+}
+
+interface Schema extends GraphQLSchema {
+  _typeMap: TypeMap;
+}
+
+interface OperationNode {
+  operation: "query" | "mutation";
+}
+
+type SchemaNode = SelectionNode | DefinitionNode;
+
+function isOperationDefinition(node: any): node is OperationNode {
+  return node.kind === Kind.OPERATION_DEFINITION;
+}
+function isFieldNode(node: SchemaNode): node is FieldNode {
+  return node.kind === Kind.FIELD;
+}
+
+// We don't treat OperationDefinitions as Definitions but as entry points for our execution
+function isDefinitionNode(node: SchemaNode): node is DefinitionNode {
+  return (
+    node.kind === Kind.FRAGMENT_DEFINITION ||
+    isTypeSystemDefinitionNode(node) ||
+    isTypeSystemExtensionNode(node)
+  );
+}
+
+interface FieldWithResolver extends GraphQLField<any, any> {
+  resolve: GraphQLFieldResolver<any, any, any>;
+}
+
+function isFieldWithResolver(
+  field: GraphQLField<any, any>
+): field is FieldWithResolver {
+  return field.resolve instanceof Function;
+}
+
+export function graphqlObservable<T = object>(
+  doc: DocumentNode,
+  schema: Schema,
+  context: object
+): Observable<{ data: T }> {
+  if (doc.definitions.length !== 1) {
+    return throwObservable("document root must have a single definition");
+  }
+  const types = schema._typeMap;
+
+  return resolve(doc.definitions[0], context, null, null).map((data: T) => ({
+    data
+  }));
+
+  function resolve(
+    definition: SchemaNode,
+    context: object,
+    parent: any,
```

Comment:
Is it possible to be a bit more specific here?

Reply:
Unfortunately, as far as I understand TS and the problem I would say no.

This can in an abstract case be any object, any array, or any primitive value.
Once we update typescript we could and should move it to the newly created `unknown` type, but for now any is all there is.

To be more precise and work with generics I think we would need to have the schema at compile time somehow assembled into TS, but also then there are a lot of questions to be answered.